### ISSUE TITLE
[css-text-3]  Segment Break Transformation Rules for East Asian Width property of A

### ISSUE THREAD

**Issue by hax** [w3c/csswg-drafts#337]
https://drafts.csswg.org/css-text-3/#line-break-transform

> Otherwise, if the East Asian Width property [UAX11] of both the character before and after the line feed is F, W, or H (not A), and neither side is Hangul, then the segment break is removed.

As this rule, common use cases of quotation marks in Chinese

```
简体中文的
“引号”
两边不应该有空格。
```

will have unexpected spaces, because quotation marks are _A_.

Ideally, we should consider the language information of the context. If the context is East Asian language, _A_ should be treat as _W_. Even in the unknown language context, if any side of the line feed is _A_ and other side is _F_, _W_ or _H_, the segment break should also be removed.

**Comment by fantasai** [comment]
My concerns here are:
* Removing spaces where they currently aren't can break existing pages.
* The proposed behavior is more complex to understand and more complex to implement for what is a fairly low-level operation.

I'm happy to make the change if i18n recommends it and implementors agree, but I am hesitant to do so for these reasons.

**Comment by hax** [comment]
@fantasai Current segment break rule in the draft already change the traditional behavior, and up to now no browser implement it.

Do you mean you just want to drop the rule totally?

And I don't think my proposal is much complex than current.

Current rule:
if the East Asian Width property [UAX11] of both the character before and after the line feed is F, W, or H (not A)

My proposal: 
if the East Asian Width property [UAX11] of both the character before and after the line feed is F, W, H, or A, except both is A.

**Comment by astearns** [comment]
@r12a waiting on i18n feedback before we get this on the CSSWG agenda again

**Comment by fantasai** [comment]
Just to clarify, the proposal is that if lang=zh|ja|yi then A->W otherwise A->N for the purpose of line-break transformations?

I think that should probably be okay. I would be against making A->W the general case.

**Comment by hax** [comment]
@fantasai In fact there are two proposals:

1.  If the context is East Asian language, A should be treat as W.

2. If context is not available (or if proposal 1 is not accepted), modify current Segment Break Transformation Rules for A:
If one side of line break is A and other side is F/W/H (which means it's very likely in east asian context), then treat A as W.

**Comment by upsuper** [comment]
I think the motivation is reasonable, and some A should be treated as W in CJ context, especially quotations in Chinese. But I am concerned about A+A case, especially given that there are lots of letters in A.

The safest thing to do is probably this: if the context is Chinese or Japanese, and one side of line break is a punctuation in A, and the other side is F/W/H, then the segment break is removed.

**Comment by fantasai** [comment]
Proposed fix:
```
diff --git a/css-text-3/Overview.bs b/css-text-3/Overview.bs
index 5f59b62..df1320b 100644
--- a/css-text-3/Overview.bs
+++ b/css-text-3/Overview.bs
@@ -669,6 +669,12 @@ Languages and Typesetting</h4>
           the character before and after the segment break is <code>F</code>,
           <code>W</code>, or <code>H</code> (not <code>A</code>),
           and neither side is Hangul, then the segment break is removed.
+        <li>Otherwise, if the <a>East Asian Width property</a> of either
+          the character before or after the segment break is <code>A</code>,
+          and that character’s <a>content language</a> is Chinese or Japanese,
+          and the character on the other side of the segment break is
+          <code>F</code>, <code>W</code>, or <code>H</code>,
+          and neither side is Hangul, then the segment break is removed.
         <li>Otherwise, the segment break is converted to a space (U+0020).
       </ul>
 
@@ -2385,6 +2391,7 @@ Changes</h2>
     <li>For ''word-break: break-all'', switched to UAX14 notion of “letters”,
         since that handles symbols better.
     <li>Added ''line-break: anywhere''.
+    <li>Tweaked handling of Ambiguous characters during <a href="#line-break-transform">segment break transformation</a>.
   </ul>
 
   <h3 class="no-num" id="changes-2013">
```

**Comment by fantasai** [comment]
@upsuper @kojiishi @hax Checked in a fix, based on Xidorn's suggestion. A+A still keeps the space, but A+F or F+A will delete the space if the A's language context is Chinese/Japanese/Yi. This is more conservative than the original request, because we don't want to break existing pages and A+A is reasonably common on non-CJK pages. An interesting question is, should we be checking the language on the segment break instead of on the A?

**Comment by hax** [comment]
@fantasai 
In my original suggestion I also didn't think A+A should delete space if we don't know whether we are in East-Asian context. 

> An interesting question is, should we be checking the language on the segment break instead of on the A?

It's basically same as my "If the context is East-Asian language, A should be treat as W", but I believe checking language on the segment break is much more precise and clear.

**Comment by upsuper** [comment]
@fantasai I think our discussion was concluded that we do that only for punctuations in A in that language context? It doesn't seem to me other A should have that behavior.

**Comment by fantasai** [comment]
OK, switched to checking the language of the segment break (rather than the A character), and restricted that rule to punctuation only.

More fun: Unicode decided to categorize emoji as Wide for some reason. >:[

**Comment by fantasai** [comment]
Fixed to treat Emoji the same as an Ambiguous character: a6aa4d856a34b7c87f23a7215635a0b4f353f63b
Why it's not Ambiguous to begin with, I don't know.

**Comment by kojiishi** [comment]
> Fixed to treat Emoji the same as an Ambiguous character

I think Emoji is too much; it's sometimes surprising and unexpected. The data here:
http://unicode.org/Public/emoji/latest/emoji-data.txt

U+0023, U+002A, U+0030-0039 are probably not desired.

**Comment by fantasai** [comment]
@kojiishi Can you explain what you think the spec should say about this? Definitely we can't rely on EAW for emoji, they are totally inconsistent. E.g. U+1F600 Grinning Face is EAW=Wide while U+263A Smiling Face is EAW=Neutral. Our rules need to treat them the same somehow, and definitely we can't treat emoji as Wide here.

**Comment by kojiishi** [comment]
I prefer not to mention. It has historical reasons to be inconsistent afaiu, Emoji is hard because it's sometimes Emoji but sometimes is not, depends on fonts. In this case, it affects only when author inserted a segment break before or after. Also even though there might be cases where it looks strange, it's interoperable, right?

**Comment by frivoal** [comment]
To avoid the problem mentioned by @kojiishi (U+0023, U+002A, U+0030-0039), we could for this purpose treat W emoji and N emoji as A.

[N emoji](https://unicode.org/cldr/utility/list-unicodeset.jsp?a=%5B%3AEmoji%3A%5D%26%5B%3AEast_Asian_Width%3Dneutral%3A%5D&g=&i=)
[W emoji](https://unicode.org/cldr/utility/list-unicodeset.jsp?a=%5B%3AEmoji%3A%5D%26%5B%3AEast_Asian_Width%3Dwide%3A%5D&g=&i=)

The only ones in all that that don't seem to me to really be "emoji" as commonly understood by people are:
* U+00A9 COPYRIGHT SIGN
* U+203C DOUBLE EXCLAMATION MARK
* U+2049 EXCLAMATION QUESTION MARK

But even then, U+00AE REGISTERED SIGN and U+2122 TRADE MARK SIGN are A as well, so lumping COPYRIGHT SIGN with them doesn't bother me.

As for U+203C, U+2049, they both are sentence ending punctuation. In Chinese and Japanese typesetting, spaces are generally not inserted around sentence ending punctuation, so treating them as A and discarding the spaces seems OK too.

**Comment by kojiishi** [comment]
Thank you for thinking through how to avoid problems. While I understand there are ways to avoid problems and there are cases where it is helpful, I think troubles win over the benefit.

**Comment by frivoal** [comment]
Despite its classification as EAW=N, **U+263A Smiling Face** in Chrome (on macOS with default fonts) looks like <img width="23" alt="screenshot 2018-12-13 03 04 30" src="https://user-images.githubusercontent.com/113268/49889079-f9ce0700-fe83-11e8-8660-380b04f6244f.png"> if `lang=ja` and like <img width="19" alt="screenshot 2018-12-13 03 04 37" src="https://user-images.githubusercontent.com/113268/49889094-03576f00-fe84-11e8-8f8e-32281b821d66.png"> if not. That looks like the behavior you'd expected of EAW=A, so I think we're going to have to do some overriding of EAW for emoji anyway to get things to make sense.

**Comment by astearns** [comment]
@kojiishi given the discussion in the working group above, can you clarify your opinion on the change that's in the editor's draft? Are you OK with this, or would you prefer to revert it?

**Comment by fantasai** [comment]
Fwiw I would strongly disagree with leaving it as-is. Treating [☺ U+263A SMILING FACE](https://unicode.org/cldr/utility/character.jsp?a=263A&B1=Show) and [😀 U+1F600 GRINNING FACE](https://unicode.org/cldr/utility/character.jsp?a=1F600&B1=Show) differently is hostile to authors imho.

**Comment by kojiishi** [comment]
> Are you OK with this, or would you prefer to revert it?

Sorry it seems I wasn't clear enough. By "revert" do you mean there was already a resolution? I guess I missed it then.

In short, I'm not interested in implementing it and rather negative, but if other impls are interested in or if there was already a resolution, I would not object.

Longer version; IIUC this issue contains two related but separate issues; EAW=A and Emoji.

Blink currently implements segment transformation rules including CJK cases in LayoutNG, and IIUC it makes Blink interoperable with Gecko. I think that it will not get updated further for the first phase of LayoutNG.

For EAW=A,  I'm neutral on this. If other implementers ship it, I may consider in future. fantasai and I discussed this several years ago, and she was negative at that point, because it can solve some cases but never can solve all cases, leaving inconsistency. Now it looks like she wants to try to solve some cases. It's still inconsistent and rules is probably not clear to authors, but solving some non-rare cases maybe a win.

For Emoji, I'm rather negative. I'm not interested in implementing it, probably in a foreseeable future. I'm not sure how far we plan to go on this feature, but as I understand from the motivations, what we really want is the consistency for color Emoji, correct? Then what we need is the logic to determine color Emoji, because both U+263A SMILING FACE and VS16 U+0030 DIGIT ZERO should look like Emoji. We will also need to take CSS Fonts properties we define to switch between text and color Emoji. Blink currently determines the used font in the shaping phase, and we don't want to shape during the white-space collapsing.

If we were to give up at some point, it will be implementable, but it means we will accept to live with inconsistencies at some point, so the benefit is limited. The user cost still exists, since ASCII range is included, the logic for every line end runs on every page of every user, consuming loading time, batteries, etc. Current other rules can be optimized out for plane 0 (U+0000-U+00FF) content.

So the question to me is that, assuming we will accept inconsistencies at some point, is improving the inconsistency a little more worth the user cost? It may depend on where we will give up, but I prefer saving user cost, because I think authors can figure it out if the rules are interoperable across all browsers.

**Comment by kojiishi** [comment]
One more example of inconsistency, from [EastAsianWidth.txt]:
> 2295;A           # Sm         CIRCLED PLUS
> 2296..2298;N     # Sm     [3] CIRCLED MINUS..CIRCLED DIVISION SLASH

EAW is about legacy encoding. As Ken suggested, when we start using EAW=A in whitespace collapsing, or maybe even W too, we will need to live with some level of inconsistencies.

I mean, I'm still fine whether we use EAW=A or not to use EAW=A, each has improved cases and each has additional inconsistencies. Haven't investigated which is net plus. But I wish us not to try more.

[EastAsianWidth.txt]: https://www.unicode.org/Public/UNIDATA/EastAsianWidth.txt

**Comment by litherum** [comment]
> As Ken suggested

Who is Ken and can where can we read their suggestion?

**Comment by kojiishi** [comment]
> Who is Ken and can where can we read their suggestion?

Sorry, Ken is the editor of the Unicode East Asian Width spec, and I thought I read fantasai's comment saying she contacted the editor, but can't find now. Probably I'm confused with some other issues, please disregard that part.

**Comment by litherum** [comment]
The spec says "browser implementations do not currently follow these rules". From my testing, Chrome, WebKit, and Edge don't seem to follow these rules at all, whereas Firefox seems to follow some of them.

The entire section smells like a heuristic, and given the lack of implementations, and the disagreement in this thread, I don't think we (collectively, the Web community) are ready to enshrine a particular behavior in the spec. Right now, the spec seems to describe a particular behavior that solves a collection of problems, but without more implementor experimentation, I don't have much confidence that this particular solution is the best solution.

I believe we can, however, agree on the intent of this section (to keep scripts which don't use spaces free from those harmful spaces), so perhaps the spec should be worded that way, rather than prescribing a particular solution just yet.

**Comment by fantasai** [comment]
@litherum The spec had a definite heuristic using EAW, which is afaik, what Firefox implemented. This issue was opened asking for spaces to disappear in more cases, specifically when one side was ambiguous in cases where the content language was known to be CJY. The CSSWG discussed and resolved on that in https://lists.w3.org/Archives/Public/www-style/2016Oct/0068.html

I unfortunately tagged the Emoji discrepency against this issue in https://github.com/w3c/csswg-drafts/issues/337#issuecomment-379776875 ; it's actually a separate issue, that the various wingdings have different behavior because the EAW is not consistently assigned, and thus our transformation rules are weirdly inconsistent from an author's point of view. @frivoal proposed a solution to this by tailoring the EAW of Emoji as he describes in https://github.com/w3c/csswg-drafts/issues/337#issuecomment-444316214

I'm not entirely sure which part @kojiishi is objecting to. I'm OK if our space-collapsing heuristics are more limited than requested in the original issue, but having Emoji behave inconsistently is really weird and imho not acceptable. https://github.com/w3c/csswg-drafts/issues/337#issuecomment-446775351

**Comment by kojiishi** [comment]
> I'm not entirely sure which part kojiishi is objecting to.

Because I didn't object!! Jokes aside, I really didn't. Let me try again.

1. I'm neutral on EAW=A issue. Need more investigation and experiments.
2. I'm rather negative on Emoji issue.
3. Either way, I will not object if other impls are interested in, or WG has consensus.

Implementation-wise:

1. Blink has implemented the spec one year or so ago in LayoutNG.
2. May consider EAW=A after we ship LayoutNG.
3. Emoji has some technical difficulties that we may not implement.

Please let me know if this is not still clear, I'll increase the time at English class.

**Comment by litherum** [comment]
Koji said:
> I'm not sure how far we plan to go on this feature

And this is more-or-less what I’m saying, too.
- Having a simple heuristic that doesn’t work well would be unfortunate
- Adding more and more special cases on top of a heuristic that didn’t work well in the first place would just end up being a pile of hacks
- Having a super complicated heuristic would be a headache for both authors and implementors

Instead, of somehow trying to synthesize whitespace collapsing rules from the East_Asian_Width property and the Emoji property, we should allow for more experimentation to try to come up with a more elegant solution (perhaps in the Unicode consortium). We have an opportunity now because there isn’t compact risk.

**Comment by litherum** [comment]
> That convo

I'd really like to hear what Ken has to say on this topic. If Ken is Ken Lunde, the editor of the East Asian Width property, his feedback would be very valuable.

**Comment by kenlunde** [comment]
I felt the hairs on the back of my neck tingle. 🥃

I recently conveyed to @fantasai and @frivoal in a private exchange that the UTC is extremely reluctant to make changes to EAW, and the latest substantive change was to add a note at the end of [Section 2, _Scope_](https://www.unicode.org/reports/tr11/#Scope), in hopes that it would discourage change requests:

> The East_Asian_Width property is not intended for use by modern terminal emulators without appropriate tailoring on a case-by-case basis. Such terminal emulators need a way to resolve the halfwidth/fullwidth dichotomy that is necessary for such environments, but the East_Asian_Width property does not provide an off-the-shelf solution for all situations. The growing repertoire of the Unicode Standard has long exceeded the bounds of East Asian legacy character encodings, and terminal emulations often need to be customized to support edge cases and for changes in typographical behavior over time.

What is being discussed here is sufficiently different than what is conveyed in the note that is quoted above, but the premise remains the same.

Prior to that, characters that have the _Emoji_Presentation_ property were changed to EAW=W, along with a note about treating _emoji presentation sequences_ as EAW=W (because the property deals in characters, not sequences). Keep in mind that characters that fall into _emoji presentation sequences_ are ambiguous as to their emoji presentation, and require (according to Unicode) an explicit Variation Selector, VS16 (U+FE0F), to indicate emoji presentation. Without an explicit Variation Selector, the EAW property value for such characters is ambiguous without drawing on one or more other properties, or through tailoring.

EAW is about resolving character width as a binary condition—sometimes necessarily via tailoring—in terms of whether to treat a particular character as half-width or full-width in the context of East Asian text processing. It really has nothing to do with the treatment of spaces, which makes me feel very uneasy about the use of this property in such a context. Ignoring the 800-pound gorilla that is represented by the CJK Unified Ideographs blocks, an extraordinary large number of characters are completely outside the scope of East Asian text, either because they are completely unrelated scripts or fall outside of the half-width/full-width paradigm.

Two other statements in the same section of EAW should be considered (emphasis mine):

> It does not provide rules or specifications of how this property might be used in font design or **line layout**, because, while a useful property for this purpose, it is only one of several character properties that would need to be considered.

> Instead, the guidelines on use of this property should be considered recommendations based on a particular legacy practice **that may be overridden by implementations as necessary**.

Anyway, I have asked a couple of Unicode experts who are better versed than me about properties related to segmentation and line breaking in hopes that they can offer solutions that don't involve EAW. The holidays are upon us, so there is likely to be delays in getting substantive or helpful feedback.

**Comment by kojiishi** [comment]
Thank you Ken, this is really helpful and great start, and thank you @litherum for involving Ken into the discussion.

I think what we need to agree first is how much accuracy we expect in this feature. No matter what we do, we will not get 100% accuracy. Even if we define a new Unicode property, the property cannot provide the perfect result because of unifications, Emoji presentations, etc. Remember the discussion on the vertical orientation property, it took over 2 years to define but we knew we need a CSS property to override because it will never be perfect.

The strategy I prefer is to pick one existing Unicode property, and live with inconsistencies the property cannot distinguish well. It is easy to explain, author can learn what are the common inconsistencies, fast to compute, easy to implement, and authors are unlikely to be troubled by implementation bugs.

It is very tempted to include a few characters, such as quotation marks as reported in this issue, but that is likely to start an endless discussion whether to include another character or not, and often it is ambiguous due to the nature of Unicode.

If the strategy works for us, I still think EAW=F/W does a reasonable job, but if people disagrees, other properties that are more appropriate to use in line layout can provide different combination of consistencies and inconsistencies.

Note, given the last discussion, in order for the discussion not be constrained by existing content, I'm thinking to disable it in LayoutNG until we get better consensus.

**Comment by frivoal** [comment]
> I have asked a couple of Unicode experts who are better versed than me about properties related to segmentation and line breaking

This is **not** about line breaking though. This is about processing U+000A that are present in the source code of a document, to decide if a space should be inserted in its place (as would be appropriate for English, where words are space separated), or not (as would be appropriate for Chinese or Japanese, where they are not). This is not about line breaking in the rendered layout.

Finding a rule that would work for all languages in all situations is unrealistic, and is not needed because authors can just avoid using line breaks in their source code to guard against such space insertion. However, finding a subset where we can reliably determine that inserting a space would be the wrong thing to do enables authors of space-less languages to format their source code freely in more situations, and enjoy some of the benefits that users of space-separated languages already enjoy.

Due the languages from which EAW=F/W/H characters come, I believe the proposed rules safely identify such a subset.

By safely, I mean "will not fail to insert a space where one is expected".

> I still think EAW=F/W does a reasonable job, but if people disagrees[...]

I agree that this rule (the second bullet point in [section 4.1.2](https://drafts.csswg.org/css-text-3/#line-break-transform)) does a reasonable job indeed. I won't die on this hill if the consensus is that doing more than that is overkill.

I do think we could do a better job, and that the additional rule about EAW=A (the third bullet point) brings significant benefits (see the situation discussed in the initial comment on this issue) at moderate cost. I'd prefer if we did it. I won't object if we don't.

Whether we do that part or not, Emoji will be inconsistent anyway, and I think it would be nice to fix as that's annoying, and I believe the proposed tailoring to do so is safe. But I'm ok if that's where you want to draw the line.

To me, these are the "appropriate tailoring" that UAX11 is calling for. 

I would not be opposed to using some other unicode property (or combination of properties) than EAW (with the proposed tailoring) if we had a realistic candidate of another classification that works for reliably identifying a safe subset, but I don't think it is helpful to drop EAW based rules for either of the following reasons:
* Even if it provides the information we want, it wasn't designed for that.
* Even if none exist, we can imagine that there should be a theoretical classification that would cover a larger safe subset.

If the result is wrong (as in "inserts spaces where there shouldn't be"), or if there's an easier way to get an equally good (or better) result, then sure. But let's keep in mind [the priority of constituencies](https://www.w3.org/TR/html-design-principles/#priority-of-constituencies):
> consider [...] authors [...] over theoretical purity

----
Digression:

> EAW is about resolving character width as a binary condition [...] half-width or full-width

Even if its goal is to classify things between narrow and wide, given that EAW is a property with 6 values (F/H/W/Na/A/N), describing it as binary undersell the amount of information it caries.

**Comment by frivoal** [comment]
Here's an attempt at stating the problem independently of the solution:

* We want to identify cases where the fact that a line break is allowed, as evidenced by the fact that one was used in the source code, does not imply that a space would be acceptable in the rendered document if the text was wrapped elsewhere (or not at all).

* A rule that only finds a subset of cases where this is true is both acceptable and unavoidable. Acceptable because authors can refrain from using line breaks in the source code even if that's inconvenient, unavoidable because of too much linguistic and typographic diversity and too many corner cases.

* On the other hand, false positives *must* be avoided, as it would likely break existing pages.

* This is only useful to authors to the extent that it is consistent across browsers, as otherwise the result would be unwanted spaces in some browsers.

* The results *should* be unsurprising, otherwise authors risk shying away from using it.

* To the extent it relies on classifying characters, it *must* do so based on properties and classifications maintained by Unicode (or combination thereof), as the CSS-WG cannot maintain character by character classifications while Unicode continues to expand.

* It *may* rely on things other than classifying characters, such as the content language.

* It *must* be practical to implement performantly.

----

On the last point, I trust implementors' judgement better than mine. But that aside, I believe each of the proposed rules in [section 4.1.2](https://drafts.csswg.org/css-text-3/#line-break-transform) satisfy all these criteria.

I also don't think I've heard any other proposal (except adopting a subset of these rules) that does.

**Comment by kenlunde** [comment]
@frivoal What you wrote just above is helpful. In an effort to confirm my understanding, please confirm that the following are accurate and coherent statements:

_The intent is to un-break lines that are not marked with an explicit line-break (U+000A and friends), and which may have one or more intervening white-space characters (such as spaces or tabs) that serve as indentation for easier source viewing. The problem is determining whether to altogether remove the line-break and any intervening white-space characters, or to replace them with a space (U+0020)._

**Comment by fantasai** [comment]
@kenlunde s/that are not marked/that are marked/ but otherwise yes. (If there is no line feed, then all spaces and tabs are collapsed to a space, they never disappear. If there is a line feed, then we strip any adjacent spaces and tabs and then contemplate whether to remove the line feed or leave behind a space.)

**Comment by fantasai** [comment]
The purpose of this is indeed to unbreak lines of text that are broken in the source code.

**Comment by kenlunde** [comment]
Right. I edited my statement above just now to remove "not."

Also, with regard to the Korean problem, or its special-casing, I think that you can safely assume that a "space" can be used to un-break such lines. While Korean text can break on spaces **and** between hangul syllables, I see no reason why such lines would break on anything other than spaces in this context. Of course, you may have already considered this, but I am conveying the above suggestion on the off chance that you didn't.

**Comment by faceless2** [comment]
For what it's worth we've just implemented this as currently specified, and it makes a real mess of some tests, e.g. [CSS2/generated-content/content-counter-004-ref.xht](https://github.com/web-platform-tests/wpt/blob/master/css/CSS2/generated-content/content-counter-004-ref.xht) - spaces between U+25FE (black square) are removed, due to the EAW property being "W".

I think basing the decision on Unicode Block rather than EAW property is certainly the way to go. In an effort to roll this forward I've had a hunt through the Unicode blocks and come up with a list that could be used as a starting point:

<a href="https://en.wikipedia.org/wiki/CJK_Radicals_Supplement_(Unicode_block)">CJK Radicals Supplement</a>
<a href="https://en.wikipedia.org/wiki/Kangxi_Radicals#Unicode">Kangxi Radicals</a>
<a href="https://en.wikipedia.org/wiki/Ideographic_Description_Characters_(Unicode_block)">Ideographic Description Characters</a>
<a href="https://en.wikipedia.org/wiki/CJK_Symbols_and_Punctuation">CJK Symbols and Punctuation</a>
<a href="https://en.wikipedia.org/wiki/Hiragana_(Unicode_block)">Hiragana</a>
<a href="https://en.wikipedia.org/wiki/Katakana_(Unicode_block)">Katakana</a>
<a href="https://en.wikipedia.org/wiki/Bopomofo_(Unicode_block)">Bopomofo</a>
<a href="https://en.wikipedia.org/wiki/Kanbun_(Unicode_block)">Kanbun</a>
<a href="https://en.wikipedia.org/wiki/Bopomofo_Extended">Bopomofo Extended</a>
<a href="https://en.wikipedia.org/wiki/CJK_Strokes_(Unicode_block)">CJK Strokes</a>
<a href="https://en.wikipedia.org/wiki/Katakana_Phonetic_Extensions">Katakana Phonetic Extensions</a>
<a href="https://en.wikipedia.org/wiki/Enclosed_CJK_Letters_and_Months">Enclosed CJK Letters and Months</a>
<a href="https://en.wikipedia.org/wiki/CJK_Compatibility">CJK Compatibility</a>
<a href="https://en.wikipedia.org/wiki/CJK_Unified_Ideographs_Extension_A">CJK Unified Ideographs Extension A</a>
<a href="https://en.wikipedia.org/wiki/Yijing_Hexagram_Symbols_(Unicode_block)">Yijing Hexagram Symbols</a>
<a href="https://en.wikipedia.org/wiki/CJK_Unified_Ideographs_(Unicode_block)">CJK Unified Ideographs</a>
<a href="https://en.wikipedia.org/wiki/Yi_Syllables_(Unicode_block)">Yi Syllables</a>
<a href="https://en.wikipedia.org/wiki/Yi_Radicals_(Unicode_block)">Yi Radicals</a>
<a href="https://en.wikipedia.org/wiki/CJK_Compatibility_Ideographs">CJK Compatibility Ideographs</a>
<a href="https://en.wikipedia.org/wiki/Vertical_Forms">Vertical Forms</a>
<a href="https://en.wikipedia.org/wiki/CJK_Compatibility_Forms">CJK Compatibility Forms</a>
<a href="https://en.wikipedia.org/wiki/Small_Form_Variants_(Unicode_block)">Small Form Variants</a>
<a href="https://en.wikipedia.org/wiki/Halfwidth_and_Fullwidth_Forms_(Unicode_block)">Halfwidth and Fullwidth Forms</a>
<a href="https://en.wikipedia.org/wiki/Kana_Supplement">Kana Supplement</a>
<a href="https://en.wikipedia.org/wiki/Kana_Extended-A">Kana Extended-A</a>
<a href="https://en.wikipedia.org/wiki/Small_Kana_Extension">Small Kana Extension</a>
<a href="https://en.wikipedia.org/wiki/Taixuanjing">Tai Xuan Jing Symbols</a>
<a href="https://en.wikipedia.org/wiki/Counting_Rod_Numerals_(Unicode_block)">Counting Rod Numerals</a>
<a href="https://en.wikipedia.org/wiki/Enclosed_Ideographic_Supplement">Enclosed Ideographic Supplement</a>
<a href="https://en.wikipedia.org/wiki/CJK_Unified_Ideographs_Extension_B">CJK Unified Ideographs Extension B</a>
<a href="https://en.wikipedia.org/wiki/CJK_Unified_Ideographs_Extension_C">CJK Unified Ideographs Extension C</a>
<a href="https://en.wikipedia.org/wiki/CJK_Unified_Ideographs_Extension_D">CJK Unified Ideographs Extension D</a>
<a href="https://en.wikipedia.org/wiki/CJK_Unified_Ideographs_Extension_E">CJK Unified Ideographs Extension E</a>
<a href="https://en.wikipedia.org/wiki/CJK_Unified_Ideographs_Extension_F">CJK Unified Ideographs Extension F</a>
<a href="https://en.wikipedia.org/wiki/CJK_Compatibility_Ideographs_Supplement">CJK Compatibility Ideographs Supplement</a>

However, this process leads me to think that we're still going go have to distinguish based on Script as well - some of these blocks will be used with Hangul:

<a href="https://en.wikipedia.org/wiki/CJK_Symbols_and_Punctuation">CJK Symbols and Punctuation</a>
<a href="https://en.wikipedia.org/wiki/Enclosed_CJK_Letters_and_Months">Enclosed CJK Letters and Months</a>
<a href="https://en.wikipedia.org/wiki/Small_Form_Variants_(Unicode_block)">Small Form Variants</a>
<a href="https://en.wikipedia.org/wiki/Halfwidth_and_Fullwidth_Forms_(Unicode_block)">Halfwidth and Fullwidth Forms</a>
<a href="https://en.wikipedia.org/wiki/Vertical_Forms">Vertical Forms</a>

and some, eg:

<a href="https://en.wikipedia.org/wiki/Yijing_Hexagram_Symbols_(Unicode_block)">Yijing Hexagram Symbols</a>
<a href="https://en.wikipedia.org/wiki/Taixuanjing">Tai Xuan Jing Symbols</a>
<a href="https://en.wikipedia.org/wiki/Counting_Rod_Numerals_(Unicode_block)">Counting Rod Numerals</a>

are likely to be used in any script.

Not doing any segment break transformation _unless_ the script is Chinese, Japanese or Yi is going to limit the impact of any side effects.

**Comment by faceless2** [comment]
We need to add <a href="https://www.unicode.org/charts/PDF/Unicode-13.0/U130-30000.pdf">CJK Unified Ideographs Extension G</a> to that list, which is in the upcoming Unicode 13.0.

I'm also told that 99.9% of Korean text that you will find on the web uses Western (aka ASCII) punctuation, so perhaps not as much as an issue as first thought, at least in practical terms.

**Comment by faceless2** [comment]
The "ghost of christmas past", who whispered in my ear about CJK Unified Ideographs Extension G, also suggests that anything Lisu and Khitan Small Script, the latter of which is new in Unicode Version 13.0, should be on the list. 👻

**Comment by kojiishi** [comment]
Thank you Mike for the investigation, this is really helpful. [This code snippet in WebKit](https://trac.webkit.org/browser/webkit/trunk/Source/WebCore/platform/graphics/FontCascade.cpp#L1239) may help developing the list too.

Maybe we should agree on the expected accuracy first. My basic idea is:
1. I do not expect the perfectly accuracy in all cases. It is a heuristic algorithm, which can't be perfect  anyway. This is ok, as long as browsers are interoperable, it should not be too hard to find the failures of the heuristic algorithm.
2. I would like this algorithm to be simple and fast. This logic is a bit "hot"; runs on every source line for all pages including English pages, consumes battery and CPU cache spaces for all users.
3. In my opinion, I think we should avoid excessive removal for non-CJK scripts. This means that when ambiguous, choose not to remove. Not ideal for CJK authors, and may not look consistent in some cases, but still a great improvement, and authors can learn as long as implementations are interoperable.
4. The basic philosophy we applied for `text-orientation` is to use one Unicode property, or ask Unicode to add a new property. If you found other properties than the Unicode Block that can produce better results, I don't insist on the Unicode Block at all, but I prefer to stick on using one property.

Do these look reasonable? Any opinions, additions, or change suggestions?

**Comment by kojiishi** [comment]
The [VerticalOrientation](https://www.unicode.org/Public/UNIDATA/VerticalOrientation.txt) property could be a good input too.

**Comment by kojiishi** [comment]
A wild idea came up, maybe the [VerticalOrientation](https://www.unicode.org/Public/UNIDATA/VerticalOrientation.txt) property is better than the Unicode Block for this purpose, rather than just using it as a reference to develop a list of blocks?

**Comment by faceless2** [comment]
I presume the reason we're discussing heuristics at all, and not simply adding another value to `text-space-collapse` in [css-text-4](https://drafts.csswg.org/css-text-4/#propdef-text-space-collapse) which says "collapse segment breaks to nothing" is because (like https://github.com/w3c/csswg-drafts/issues/4576) this behaviour is supposed to be context-dependent - i.e. it depends on the characters on either side?

And the intention is, roughly, if the segment break is between two CJK ideographs, or between a CJK ideograph and punctuation, collapse it to nothing?

I ask because - if this behaviour must remain context-dependent - I'm wondering if it would be easier to add a value to `text-space-collapse` to turn this behaviour on, and instead list the contexts where it _wouldn't_ apply. So make the rule "if this flag is on, collapse all segment breaks _except_ those either side of characters of class (AL|HL|SA)".

Alternatively: it looks like Gecko is currently collapsing segment breaks between two ideographs, but not between ideographs and punctuation, and Blink/Webkit are doing neither. Perhaps segment breaks could always collapse between two ideographs (an easy and unambiguous test for UAX#14 class ID), but only collapse according to the more complex heuristics if the appropriate property was set? In other words, "always collapse segment breaks between ID characters. Only collapse other (some? all?) segment breaks if `text-space-collapse: collapse-break` is set"

(both of these are attempts to reduce both the processing cost of evaluating the heuristic, and the cost of getting the heuristics wrong).

**Comment by kojiishi** [comment]
I'm ok to add a new value, but what are the benefits of the new value? Is it to prevent regressing non-CJK content? How is it different from choosing conservative heuristics?

> A wild idea came up, maybe the VerticalOrientation property is better...

I take this back. I remember VerticalOrientation is still too aggressive.

**Comment by faceless2** [comment]
> Is it to prevent regressing non-CJK content?

Yes, exactly.

> How is it different from choosing conservative heuristics?

To my non-expert eyes, this particular heuristic appears to be quite hard to get right. That's purely based on the discussion in La Coruña, and re-reading all the comments on this issue. So I figured it's worth exploring if there's a way to remove the heuristic, or at least drastically reduce its scope.

**Comment by fantasai** [comment]
> For what it's worth we've just implemented this as currently specified, and it makes a real mess of some tests, e.g. CSS2/generated-content/content-counter-004-ref.xht - spaces between U+25FE (black square) are removed, due to the EAW property being "W".

This is because Unicode changed the EAW of a lot of characters in an effectively random and backwards-incompatible way when it introduced Emoji. The results based on e.g. [Unicode 6](https://www.unicode.org/Public/6.1.0/ucd/EastAsianWidth.txt), when these rules were written, would have been quite sensible. :/ Trying to compensate for this change is one of the reasons the rules became too complicated...

I've committed an initial draft of the Unicode block-based approach. I think the interesting questions remaining are:
- Bopomofo
- Yijing Hexagram Symbols / Tai Xuan Jing Symbols / Counting Rod Numerals
- Enclosed ideographics

I'm leaning towards yes on enclosed ideographics, no on the symbols, and I don't know enough about Bopomofo when it is used as a stand-alone script to say.

Lisu and Khitan both use spaces; they should not therefore discard them during collapsing. Small forms etc. are primarily used with Chinese and Japanese, not Korean, so I think it's reasonable to include them here. (Keep in mind also that both sides of the break need to belong to the set in order to discard, and Hangul is excluded.)

**Comment by frivoal** [comment]
> Lisu and Khitan both use spaces

Are we talking about one or both of these Khitan (presumably the former, as I don't think the later is in Unicode):

https://en.wikipedia.org/wiki/Khitan_small_script
https://en.wikipedia.org/wiki/Khitan_large_script

If yes, do they really use spaces? Where can I learn more about that?

If not, what are we talking about?

**Comment by fantasai** [comment]
@frivoal https://www.unicode.org/versions/Unicode13.0.0/ch18.pdf

**Comment by fantasai** [comment]
The WG resolution to switch to Unicode Blocks has been edited in. I opened up https://github.com/w3c/csswg-drafts/issues/4993 and https://github.com/w3c/csswg-drafts/issues/4993 as follow-up issues. Closing out discussion here, since we've veered pretty far off the original topic.

**Comment by fantasai** [comment]
Refiled the OP as https://github.com/w3c/csswg-drafts/issues/5017
